cmd/rpcdaemon/commands/trace_api.go

```diff
@@ -0,0 +1,246 @@
+package commands
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/ledgerwatch/turbo-geth/common"
+	"github.com/ledgerwatch/turbo-geth/common/dbutils"
+	"github.com/ledgerwatch/turbo-geth/common/hexutil"
+	"github.com/ledgerwatch/turbo-geth/core/rawdb"
+	"github.com/ledgerwatch/turbo-geth/eth"
+	"github.com/ledgerwatch/turbo-geth/ethdb"
+	"github.com/ledgerwatch/turbo-geth/turbo/adapter"
+	"github.com/ledgerwatch/turbo-geth/turbo/transactions"
+)
+
+// TraceFilterRequest represents the arguments for trace_filter.
+type TraceFilterRequest struct {
+	FromBlock   *hexutil.Uint64   `json:"fromBlock"`
+	ToBlock     *hexutil.Uint64   `json:"toBlock"`
+	FromAddress []*common.Address `json:"fromAddress"`
+	ToAddress   []*common.Address `json:"toAddress"`
+	After       *uint64           `json:"after"`
+	Count       *uint64           `json:"count"`
+}
+
+// TraceAPI
+type TraceAPI interface {
+	Filter(ctx context.Context, req TraceFilterRequest, config *eth.TraceConfig) ([]interface{}, error)
+}
+
+// TraceAPIImpl is implementation of the TraceAPI interface based on remote Db access
+type TraceAPIImpl struct {
+	db        ethdb.KV
+	dbReader  ethdb.Getter
+	maxTraces uint64
+}
+
+// NewPrivateDebugAPI returns PrivateDebugAPIImpl instance
+func NewTraceAPI(db ethdb.KV, dbReader ethdb.Getter, maxTraces uint64) *TraceAPIImpl {
+	return &TraceAPIImpl{
+		db:        db,
+		dbReader:  dbReader,
+		maxTraces: maxTraces,
+	}
+}
+
+func retrieveHistory(tx ethdb.Tx, addr *common.Address) ([]uint64, error) {
+	addrBytes := addr.Bytes()
+	ca := tx.Cursor(dbutils.AccountsHistoryBucket).Prefix(addrBytes)
+
+	var blockNumbers []uint64
+
+	for k, v, err := ca.First(); k != nil; k, v, err = ca.Next() {
+		if err != nil {
+			return nil, err
+		}
+		numbers, _, err := dbutils.WrapHistoryIndex(v).Decode()
```

Comment:
You can pass `toBlock` into this function and break out of the loop if you see that `numbers[len(numbers)-1] > toBlock`

Reply:
but if only the last member of `numbers` is out of `toBlock` and the other are in range, then we cut out blocks to go over. we should take `toBlock` and compare the first element and with fromBlock, compare the last one instead